### PR TITLE
Remove deprecated package "postcss-next"

### LINKED ISSUE #14052: `postcss` fails to parse css files during builds
**Describe the bug**

When the build/CI steps run `yarn run postcss` we are failing (exit code 1). This appears from the content of postcss_error.log to be due to a parse error and related to an older version of postcss included with a deprecated package postcss-cssnext.

**To Reproduce**

in a shell

```
yarn
yarn run postcss
# "command exits with error code 1"
cat postcss_error.log
```

**Expected behavior**

postcss compiles the css in public/assets/*.css instead of failing.

**Additional context**

I would expect this started after https://github.com/forem/forem/pull/12704 in February

Logging to a file instead of displaying the error message during build was done in https://github.com/forem/forem/pull/11521 (original issue is a lot of warnings about parsable css were causing build failures, however that seems to have masked more serious issues).

I think we may want to fail the build if the postcss step fails (it's not clear why this step doesn't block the build, is it important? What does it mean that it's been failing repeatedly?)

In investigating this I removed all postcss-* resolved package versions from yarn.lock  and let yarn re-resolve. I saw two packages give deprecation warnings. These are also the only two plugins configured at all in the .postcssrc.yml file.

```
[2/6] Resolving packages...

 warning postcss-cssnext@3.1.0: 'postcss-cssnext' has been deprecated in favor of 'postcss-preset-env'. Read more at https://moox.io/blog/deprecating-cssnext/

 warning postcss-smart-import@0.7.6: This project is not maintained anymore. Please use postcss-import instead.
```

I was able to make this "go away" by removing postcss-cssnext from the package.json and from the postcssrc config. I'm not sure if that's something we should or want to do.

### PR DESCRIPTION
**Pull request by djuber**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This was deprecated a few years ago, and appears to be causing issues
with a newer (postcss v8) api.

https://moox.io/blog/deprecating-cssnext
https://github.com/csstools/postcss-preset-env/issues/32

This appears to fix https://github.com/forem/forem/issues/14052 but
was tested only by observing that the postcss command no
longer fails with an error, and not that browser ready css is
generated from the input or working for users.

~It's likely that the next step is to add back
postcss-preset-env (which does not include postcss-calc, which was I
think related to the broken parses and calling the older version of
postcss internally).~ nothing to do here - postcss-preset-env was included by webpacker automatically and is already configured.

Actually, postcss-preset-env was installed, and looked like it was configured, but was not added to a .postcssrc.yml file which was overriding the "default + loaded config" behavior and causing this to be skipped after I removed the postcss-cssnext plugin from the rc file. That's been changed in this PR (the config options from the postcss.config.js file were not brought into the yaml rc file) and once that was enabled I could confirm the postcss processor was running again (files were being parsed, and failing). 

A change to the max-width property of the runtime-banner element (used to suggest the native app to ios users) appears to have been the cause of the parse error, and an equivalent formulation has been used instead. I still don't understand what, between sass and postcss processing, was happening between calc(min(x,y)) to cause the error, but we don't do that anymore.

The error redirection is left disabled, I currently see two warnings (which are generated in a single file context but don't give the file being processed, it's not very usable but we can let it grow until it's a pain point again). 

Sure wish it told me what the file with 70K lines with the "start" value was.

```
 yarn run v1.22.5
 $ postcss public/assets/*.css -d public/assets

70881:3 ⚠  start value has mixed support, consider using flex-start instead [postcss-preset-env]                                                                      
71131:3 ⚠  start value has mixed support, consider using flex-start instead [postcss-preset-env]                                                                      

 Done in 4.47s.
```

## Related Tickets & Documents

fixes #14052

## QA Instructions, Screenshots, Recordings

The postcss script is called as a post precompile assets hook from rake, or separately executable from yarn.
The simplest proof of concept would be to run `yarn postcss`.

I don't know how to check that features supported or intended to be supported by the processor (which was erroring every time I looked at it) are actually being processed. I assume that's something like validating css loads correctly in some target set of browsers.

## Questions for reviewers

Do we want to remove the .postcssrc.yml file (and let the postcss.config.js file take effect)? 
Do we instead want to copy any of the config for postcss-preset-env into the postcssrc.yml file (options hash for plugin)?
Does the width/max-width correctly capture Fernando's intent for the `.runtime-banner`?
Are we comfortable letting the warnings show up again in the builds (the original issue leading to hiding them appears to have been a wall of warnings causing build failures due to overflow or timeouts).

### UI accessibility concerns?

Very possibly - if our accessibility was based on the css - however from what I can tell we're just removing broken software rather than changing some configuration that had been working correctly. Since it was just failing (from what I can see) I don't think we're any worse off.

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: build config change
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [x] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

Probably spot check with browsershots or some other cross-platform test tool to make sure forem isn't failing to render correctly. 

## [optional] What gif best describes this PR or how it makes you feel?

delete key

### CODE REVIEW

**Comment by djuber** [comment]
I'm feeling a little better about this - checking in the [postcss.config.js](https://github.com/forem/forem/blob/main/postcss.config.js#L6) we _already_ had been using postcss-preset-env and checking the yarn.lock file it was already installed - as a dependency of [`@rails/webpacker@5.4.0`](https://github.com/forem/forem/blob/main/yarn.lock#L1771-L1814)

This _might_ be safe to merge unless the reduce calc parts were something critical for us. I don't know how to test this.

**Comment by citizen428** [comment]
I choose you, Nick-achu! 

image

Bad puns aside, maybe @nickytonline can help you get some clarity here @djuber 😃

**Review by jacobherrington — APPROVED** [review]
:hocho:

**Comment by djuber** [comment]
I had been referencing the postcss.config.js file as the implemented configuration (it may be). As an experiment I copied that plugin's config into .postcssrc.yml (which is where the postcss-cssnext package had been configured previously) and I started to see the same kinds of configuration issues (with a postcss version conflict and a parse error)

```
djuber@forem:~/src/forem$ yarn postcss
 yarn run v1.22.5
 $ postcss public/assets/*.css -d public/assets
Unknown error from PostCSS plugin. Your current PostCSS version is 8.3.3, but postcss-preset-env uses 7.0.35. Perhaps this is the source of the error below.
Unknown error from PostCSS plugin. Your current PostCSS version is 8.3.3, but postcss-preset-env uses 7.0.35. Perhaps this is the source of the error below.
Unknown error from PostCSS plugin. Your current PostCSS version is 8.3.3, but postcss-preset-env uses 7.0.35. Perhaps this is the source of the error below.
Unknown error from PostCSS plugin. Your current PostCSS version is 8.3.3, but postcss-preset-env uses 7.0.35. Perhaps this is the source of the error below.
ParserError: Syntax Error at line: 1, column 15
    at /home/djuber/src/forem/public/assets/minimal-0f45bc13062380b41982ca2295149b59d3aec13cc69cd01db5a5ef367147ac29.css:1:223985
    at Parser.error (/home/djuber/src/forem/node_modules/postcss-values-parser/lib/parser.js:127:11)                                                                     
    at Parser.operator (/home/djuber/src/forem/node_modules/postcss-values-parser/lib/parser.js:183:18)                                                                     
    at Parser.parseTokens (/home/djuber/src/forem/node_modules/postcss-values-parser/lib/parser.js:245:14)                                                                     
    at Parser.loop (/home/djuber/src/forem/node_modules/postcss-values-parser/lib/parser.js:132:12)                                                                     
    at Parser.parse (/home/djuber/src/forem/node_modules/postcss-values-parser/lib/parser.js:51:17)                                                                      
    at parse (/home/djuber/src/forem/node_modules/postcss-custom-properties/index.cjs.js:47:30)                                                                       
    at /home/djuber/src/forem/node_modules/postcss-custom-properties/index.cjs.js:333:24
    at /home/djuber/src/forem/node_modules/postcss/lib/container.js:72:18
    at /home/djuber/src/forem/node_modules/postcss/lib/container.js:55:18
    at Rule.each (/home/djuber/src/forem/node_modules/postcss/lib/container.js:41:16)
    at Rule.walk (/home/djuber/src/forem/node_modules/postcss/lib/container.js:52:17)
    at /home/djuber/src/forem/node_modules/postcss/lib/container.js:60:24
    at Root.each (/home/djuber/src/forem/node_modules/postcss/lib/container.js:41:16)
    at Root.walk (/home/djuber/src/forem/node_modules/postcss/lib/container.js:52:17)
    at Root.walkDecls (/home/djuber/src/forem/node_modules/postcss/lib/container.js:70:19)
    at transformProperties (/home/djuber/src/forem/node_modules/postcss-custom-properties/index.cjs.js:330:8) {                                                                     
  postcssNode: Declaration {
    raws: { before: '', between: ':' },
    type: 'decl',
    parent: Rule {
      raws: [Object],
      type: 'rule',
      nodes: [Array],
      parent: [Root],
      source: [Object],
      selector: '.runtime-banner',
      lastEach: 1,
      indexes: [Object],
      [Symbol(isClean)]: false
    },
    source: { start: [Object], input: [Input], end: [Object] },
    prop: 'max-width',
    value: 'calc(min(100% - var(--su-4) * 2, 750px))',
    [Symbol(isClean)]: false
  }
}
 error Command failed with exit code 1.
```

It's possible we're _still_ not doing anything? 

When I removed the .postcssrc.yml file (only keeping the postcss.config.js unaltered) I start to see this error message as well when I run `yarn postcss` - it looks like the postcss.config.js file may have been superseded by the confiuguration in .postcssrc.yml (and doing nothing)?

File: app/assets/stylesheets/runtime-banner.scss [hunk 1]
```diff
@@ -12,7 +12,8 @@
   flex-flow: row nowrap;
   align-items: center;
   padding: var(--su-3);
-  max-width: calc(min(100% - var(--su-4) * 2, 750px));
+  max-width: 750px;
+  width: calc(100% - var(--su-4) * 2);
```
Comment (djuber) [inline app/assets/stylesheets/runtime-banner.scss:16]: @fdoxyz does this look like it captures the intended sizing from https://github.com/forem/forem/pull/13662

**Comment by djuber** [comment]
dismissing earlier review from @jacobherrington since I think this branch took a wider meander than I expected

File: app/assets/stylesheets/runtime-banner.scss (hunk 1 above)
Comment (fdoxyz) [reply app/assets/stylesheets/runtime-banner.scss:16]: I tried it out locally and didn't quite work (overflowed to the right)

IMG_F5A9BA1765D6-1

I think that the flex styling makes the banner grow to fill the container by default, that's why it only had `max-width` before. We're refactoring the way it works but the design [will remain the same for now](https://github.com/forem/rfcs/pull/249#discussion_r658085600) 🙂

Comment (fdoxyz) [reply app/assets/stylesheets/runtime-banner.scss:16]: I did see this error but CI didn't fail (the job itself). I wonder if the `min` command is the problem here though 🤔

Comment (djuber) [reply app/assets/stylesheets/runtime-banner.scss:16]: Yes, the `calc(min(...))` combination seems to be the parse problem (sass passes it through and postcss rejects it) - the real problem with this setup seems to be that postcss failing to process doesn't cause the build to fail (basically whatever processing we assumed was happening can fail silently and the build proceeds) - you get no feedback if you break something, and we expect that.

It's also possible that the ios browser's css rendering accepts this just fine and _only_ postcss's parser is problematic, that was one of the reasons I started by removing cruft from the configuration and then moved on to finding the problematic line. 

Thanks so much for spot checking, I'll see if there's some other way to handle this correctly.

**Comment by nickytonline** [comment]
I've got some other things on the go at the moment, but I can look into this when I'm back from vacation.

**Comment by djuber** [comment]
I'm going to mark this draft since there are problems introduced by the runtime banner style change

**Comment by djuber** [comment]
Closing this as stale - I think there was some information gained here (about the reason and place this is failing in `calc(min(...` ) but I'm not able to push past the change to behavior for the runtime banner.

It's possible we can (1) skip the postcss step if we don't need it, (2) remove this or another plugin (seems necessary but not sufficient), (3) change the runtime banner element to find the width another way 

One secondary goal would be to change the pipeline so that if this build step fails (as it has for months) the build fails (so we get feedback when we make breaking changes) - right now redirecting to an error log and passing over the step without requiring it to pass prevents seeing any problems without careful inspection.

**Pull request closed without merging**